### ISSUE TITLE
Add annotation/event from graph panel

### ISSUE THREAD

**Issue by Onlinehead** [grafana/grafana#1286]
Sometimes we need to add one time annotation to graphic if something irregular happen, like provider's network problems etc. That help other peoples to know what happen in that time.
But now it's not possible to add to Grafana. Automatically annotation from graphite and other data source is too hard for do it every time by hand.
What you think about adding future 'add annotation' by click on data point on graphic or by mark array and add? It can be very useful sometimes.

**Comment by jaimegago** [comment]
@torkelo With 2.x getting it's own backend this seems like it'd be easier to implement now? I think it'd be a killer feature with "ohhh" and "ahhhh" effect. I've seen commercial products demo where that "live annotation on graphs" feature was greatly advertised.
When/if you end up implementing this IMHO it'd be great to provide a backend option for the annotations other than Grafana own DB. In my case I'd like to be able to create annotation (aka events) from within Grafana that are then being stored as Graphite events. The reason is that I already use Graphite events and this way I'd have all the events in one place for consumption by different front ends (Grafana being one of them).

**Comment by addshore** [comment]
This would be really great! +1 I look forward to it!

**Comment by gopalakrishnan-subramani** [comment]
User notes, this can be stored with in dashboard model/table/document itself with attributes if that reduce the development time. The time-stamp along with user note, user name, time can be stored. 

I don't think, these dashboards shall have so many notes that affect performance. Each Graph (& Possible table) can have feature to add note, render the existing note, delete a note.

**Comment by CamJN** [comment]
+1 Would really love this

**Comment by mtanda** [comment]
+1

**Comment by torkelo** [comment]
To begin with I see this implemented as follows (some shortcut / keybinding can also be added)

Add a new action from the panel menu (hamburger, where you have Pane JSON, export, toggle Legend), the action should be “Add annotation” ,  then user can select a time range, this then opens a "Add annotation" modal, where use can specify title & description. The annotation should be saved in the Grafana built in annotation storage.

**Comment by torkelo** [comment]
right now there is no HTTP api to save a new annotation, this needs to be added in the backend, but the internal stuff is there (to save annotation etc).

https://github.com/grafana/grafana/blob/master/pkg/api/annotations.go#L9

**Comment by kylebrandt** [comment]
I'd love to have this pluggable so I can integrate creation of bosun annotations. Currently in the Bosun ui I use a brush but hold shift - that way it is easy to create the timespan of something over a graph that displays the change you want to capture.

**Comment by def324** [comment]
+1

**Comment by hraftery** [comment]
+1 this would be sensational.

It's driving me mad that every time I have to document an incident I have to jump into the command line, run influx, use the correct database, select the correct measurement, and enter a long single insert statement where every quote, every comma and every space must be precise, plus I need to convert from local time in Grafana to UTC, then to seconds since unix epoch, then add 9 zeros to convert to nanoseconds. Of course, many of those issues are influx specific, but you can see how any improvement would be a big one, and I'm sure many are in the same boat. Unless I'm missing something...?

**Comment by leehambley** [comment]
Heath, I find myself in the same position as you, such is the nature of
FOSS. I didn't have the time or relevant and immediate skill to do tribute
this feature to grafana, but I hacked together a Sinatra webapp to talk to
influx, and linked it from Grafana, it works very well, I'll comment back
here once I'm able to package it and convert the hard coded
names/credentials to environment vars so it's reusable, maybe you can use
it.

Am 01.02.2017 06:32 schrieb "Heath Raftery" <notifications@github.com>:

**Comment by hraftery** [comment]
Sounds very promising @leehambley . Coincidentally, I've just finished hacking together my own method! Didn't want to have to spend 2 hours doing it, but it's done now and much better than the manual method, so here it is for anyone to leverage:

Install the [influxdb-python](https://github.com/influxdata/influxdb-python) client. I chose Python3 but 2 would probably work just fine.

```
sudo apt-get install python3-pip
sudo pip3 install influxdb
```

Then create the script (I called it `add_annotation.py`), making sure to change the timezone to your own (DST is automatically handled):

```
#!/usr/bin/python3

from influxdb import InfluxDBClient

import pytz, datetime

def main():
	
	client = InfluxDBClient('localhost', 8086, '', '', 'annotations')
	
	title = input("Enter the title: ")
	text = input("Enter the text: ")
	tags = input("Enter the tags: ")
	time = input("Enter the local time (YYYY-MM-DD HH:MM:SS): ")

	naive = datetime.datetime.strptime(time, "%Y-%m-%d %H:%M:%S")
	my_tz = pytz.timezone("Australia/Sydney")
	local_dt = my_tz.localize(naive, is_dst=None)
	utc_dt = local_dt.astimezone(pytz.utc)
	time = utc_dt.strftime('%Y-%m-%dT%H:%M:%SZ')

	json_body = [
		{
			"measurement": "stag",
			"tags": {
			},
			"fields": {
				"title": title,
				"text": text,
				"tags": tags
			},
			"time": time
		}
	]

	client.write_points(json_body);

	print("Wrote the following point:")
	print(json_body)


if __name__ == '__main__':
    main()
```

Finally, run it!

```
chmod 755 ./add_annotation.py
./add_annotation.py
```

Fairly basic, fairly self-explanatory. No fault tolerance and the entered text cannot be scrolled through and edited (so write it in a text editor and copy&paste). But damn is it so much nicer than the manual method!

**Comment by alexanderzobnin** [comment]
@torkelo some thoughts about add annotation flow. Now it looks like you wanted:
1. Open panel menu and click _Add Annotation_ button:
![image](https://cloud.githubusercontent.com/assets/4932851/24830814/c16d9956-1c95-11e7-8b1e-6726d230f042.png)
2. Click on graph panel for selecting desired time. Then Add Annotation dialog will be opened.
3. Fill annotation info and click _Add Annotation_ button
![screenshot from 2017-04-08 19 55 10](https://cloud.githubusercontent.com/assets/4932851/24830821/e2ea9ebc-1c95-11e7-8d84-54985c6f4629.png)


I think it makes sense to notify user when he clicks _Add Annotation_ button in panel menu (1st step). I want to show tip that he should click on graph, but don't want to block flow by using modal dialog (with just this tip). Any ideas?

**Comment by torkelo** [comment]
@alexanderzobnin yea, we need some kind of hint to show that they should click or drag a region (would be cool to support region annotations as well with a start and stop). 

But focus on getting the basic feature working for now, and we can loop in @mattttt  for some ux insight . 

I have an idea for how to do the hint/tip

**Comment by torkelo** [comment]
@alexanderzobnin after discussing this with @mattttt we think the primary way to add an annotation should be to select a region on the graph directly While holding a modifier key like CTRL/SHIFT. 

We need away to educate users of this but that is a secondary problem (the Add Annotation panel menu action could open a small  modal to describe how you add an annotation by holding CTRL  and markign a region on the graph)

**Comment by alexanderzobnin** [comment]
@torkelo Ok. Initial flow in almost done, you can take a look (I can open PR, if you want). Now I'm working on backend part.

**Comment by torkelo** [comment]
Some design &  terminology decisions we have to make for annotations. 

# Terminology
- Annotation Query (what you can currently define in the annotations view in a dashboard)
- Annotation Events (The individual items rendered on graph)
- Annotation Events created from alert state change
- Annotation Events created manually by user (this issue is about this new feature) 
- Annotation Event Category (a way to tag/organize a manually created event, think `Outage`, `Deployment`, `Server upgrade`, `Marketing campaigns`).

# Question
We have 2 types for annotation events, one is `alert`, so what is the type name for the ones created manually by a user? 

# Category
The category concept will be bigger than just a way to structure an annotations as we could reuse the same abstraction to organize alert rules.  So thinking of having an category type (one for `alert` categories, and one for `manually created events` categories). The reason why this category id will be shared is that alert rules trigger alert annotation events (alert history) and would like the category to cary over from the alert rule. 

# Category vs tags
Why not organize alert rules & annotations with tags. This is to simplify the design and the queries for annotations. And will make it easier to have permissions associated with categories. But it will be a big limiting factor for how you can structure & filter annotation events & alert rules.

Thoughts?
ping @mattttt @Dieterbe

**Comment by Dieterbe** [comment]
annotate is literally defined as
> add notes to (a text or diagram) giving explanation or comment.
"documentation should be annotated with explanatory notes"
synonyms:    comment on, add notes/footnotes to, gloss, interpret, mark up
"annotate the text in Chapters 4 and 5"

( https://www.google.com/webhp?sourceid=chrome-instant&ion=1&espv=2&ie=UTF-8#q=define:annotate )

so I think it's a great fit for manually added.. things.

events is broader and could be anything. and it works for points in time as well as time ranges. so i think what's currently called annotations should actually be called events.

**Comment by hraftery** [comment]
The broader "events" over "annotations" has appeal, but there's some drawbacks: it seems I'm not alone in having one of my main measurements called "events". Real world events are inserted here with their metadata (time, signal levels, source device, type of event, etc.). I then want to "annotate" certain periods in time with comments about the state of the system or explanations of weird events. Usually they're manual and done retrospectively, but having them trigger on alert is much the same thing in my mind.

So "event" might be considered a reserved word by now, and "annotation" makes sense whether it's manual or not. So maybe `annotation alert` and `annotation entry` would suffice?

On the category vs tags front, I don't understand the issue completely but for what it's worth, the flat hierarchy of tags is appealing. Makes it easy for me to find the annotations that have been tagged "Up" or "Down" and which of the "Down" tags were associated with "Battery" or "Switch". Most appealing is that I don't have to define categories up front. I can add tags as I discover useful ways to categorise.

**Comment by openbufo** [comment]
Annotations shouldn't be limited to just text. Attaching pdf, docs, jpegs etc is preferable. In that case, users can upload documents and add to the annotation in a given time frame.

**Comment by torkelo** [comment]
@openbufo thinking linking do documents will be fine. Adding "document store" to Grafanas growing list of features feels very outside it's main focus :) 

> On the category vs tags front, I don't understand the issue completely but for what it's worth, the flat hierarchy of tags is appealing. Makes it easy for me to find the annotations that have been tagged "Up" or "Down" and which of the "Down" tags were associated with "Battery" or "Switch". Most appealing is that I don't have to define categories up front. I can add tags as I discover useful ways to categorise.

No doubt is tags more flexible and easier to use. But a hell to work with when it comes to SQL querying and filtering (and the annotation store is SQL based as they are stored in Grafana's main db). 

Here is how the latest UI looks for a single event annotation:
<img width="641" alt="screen shot 2017-04-15 at 10 53 32" src="https://cloud.githubusercontent.com/assets/10999/25062299/42da041e-21ca-11e7-9bb3-e09eed4e11c6.png">

What's nice is that you can fine tune the time and by zooming in and not loose any title / description you have written, you see the annotation and can hover over it as you edit the info. 

We also will add a way to edit / delete annotations, so hovering over one, in the annotation tooltip you will have a small edit link. That is not implemented yet though. 

The biggest challenge is how these annotations are fetched. For alert annotations they are fetched for each panel. But the plan is to use annotation queries (that you can hide if you do not want the toggle) for everything. And add the option to only show annotations on matching panels (where they are created), or on all if it's a global annotation. 

The problem we need to solve is how to add the annotation query (the thing you currently add in annotation settings) when you first add an annotation event entry, after that query is added we also need to get the user to save the dashboard.

**Comment by shiv4nsh** [comment]
We are eagerly waiting for this feature when is the release due ?

**Comment by gopalakrishnan-subramani** [comment]
Dear Torkel,

I would like to add below requests..

1. Also consider adding two additional columns, that has option to
store serialised JSON content/additional data
    that may be useful developers who are building extensions for grafana.
Since it is serialised JSON content

    Column names:
                               1. data (type string/text/blob)
                               2. data_type [describe what is stored in the
column "data"] (json, image_path, image, pdf, pdf_path etc)


     a. Developers have options for customisation of widget, include
row/widget information in the annotation
     b. Attaching document path such as pdf, jpg, or any additional notes,
geo tagging, reference variables, ranges etc.
     c. We shall not bug you for additional columns in the tables, when we
have custom features to be added :-)
     d. Reduce a lot of work for developers to create additional tables to
store those information.


    You may exclude "data" in the query result, until developers who wants
to fetch this data by
    using additional query only on need basis.

2. If you have option to include user id, that is who has added the
annotation would be helpful,
    but can be managed using request 1

3. Support markdown for the description tag.

Will you please help with request 1,   developers may use it for other
multi-purpose customisation.

Thank you,

Gopalakrishnan

**Comment by roscoe57** [comment]
any update on the delivery of this one ?
is it in the nightly builds, tempted dunno if I can wait much longer, drool....

**Comment by p2ya** [comment]
Is there any update on this and the expected release time ?

**Comment by alexanderzobnin** [comment]
This feature still in progress. It works, in general, but requires a bit more tests and UI/UX work.

**Comment by p2ya** [comment]
Thanks, is it the same as : http://docs.grafana.org/reference/annotations/
As I do not have the option available in setting.
When is the expected release time ?

**Comment by daniellee** [comment]
@p2ya Grafana 5.0.0 is the next major release and it should be available in September/October.

**Comment by roscoe57** [comment]
is this available to try in any nightly/beta/addon ?

**Comment by torkelo** [comment]
not yet, still has some ux issues we need to fix

**Comment by leehambley** [comment]
> not yet, still has some ux issues we need to fix

Documented anywhere, it'd be nice to offer some contributions if I'm able.

**Comment by roscoe57** [comment]
don't see any reference to this on 5.0.0 CHANGELOG.md
gonna cry if its bumped further

**Comment by torkelo** [comment]
After a long development process I am happy to announce that this feature is now merged and available in master & v4.6 beta (releases today),  

Checkout the docs for more info! 
http://docs.grafana.org/reference/annotations/
